xapian-core/diversify/diversify.cc

```diff
@@ -81,54 +84,63 @@ Diversify::compute_similarities()
 {
     LOGCALL_VOID(API, "Diversify::compute_similarities", NO_ARGS);
     Xapian::CosineDistance d;
-    for (unsigned int i = 0; i < points.size() - 1; ++i) {
-	for (unsigned int j = i + 1; j < points.size(); ++j) {
-	    double sim = d.similarity(points[i], points[j]);
-	    auto key = get_key(points[i], points[j]);
+    for (auto p_a : points) {
+	Xapian::docid pointid_a = p_a.first;
+	Xapian::Point point_a = p_a.second;
```

Comment:
Here (and also just below) you're copying the `Xapian::Point` object, and that's unlikely to be a cheap copy - `Xapian::Point` is a subclass of `Xapian::PointType` which has a member of type`std::unordered_map<std::string, double>` (plus another member and `Xapian::Point` adds another but those are cheaper to copy).

You can avoid the copy by making these `const` references - i.e. `const Xapian::Point& point_a = pa_second;`

Reply:
Thanks for pointing this out. Pushed a fixup commit for the same.